### PR TITLE
Introduce data-flow analysis 

### PR DESCRIPTION
**Pull request by SupunS**
Supports forward data-flow analysis (uninitialized variables)

### CODE REVIEW

**Comment by SupunS** [comment]
Do not merge yet

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticCode.java [hunk 1]
```diff
@@ -304,10 +304,13 @@
...
     ARROW_EXPRESSION_CANNOT_INFER_TYPE_FROM_LHS("arrow.expression.cannot.infer.type.from.lhs"),
     ARROW_EXPRESSION_NOT_SUPPORTED_ITERABLE_OPERATION("arrow.expression.not.supported.iterable.operation"),
-    
+
     INCOMPATIBLE_TYPE_CHECK("incompatible.type.check"),
     UNNECESSARY_CONDITION("unnecessary.condition"),
-    ;
+
+    // Dataflow analysis related error codes
+    UNITIALIZED_VARIABLE("uninitialized.variable"),
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticCode.java]: typo `UNITIALIZED` --> `UNINITIALIZED`?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticCode.java]: same for `PARTIALLY_ITIALIZED_VARIABLE`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,953 @@
+package org.wso2.ballerinalang.compiler.semantics.analyzer;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java]: Missing license header.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,953 @@
...
+    private SymbolEnv env;
+    private SymbolTable symTable;
+    private SymbolResolver symResolver;
+    private BLangDiagnosticLog dlog;
+    private Names names;
+    private Map<BSymbol, InitStatus> uninitializedVars;
+
+    private static final CompilerContext.Key<DataflowAnalyzer> DATAFLOW_ANALYZER_KEY = new CompilerContext.Key<>();
+
+    public DataflowAnalyzer(CompilerContext context) {
```
Comment (Manuri) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java]: Shouldn't this be private?

File: tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/dataflow/analysis/DataflowAnalysisTest.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import org.testng.annotations.Test;
+
+/**
+ * Test cases related to data flow analysis.
+ *
+ * @since 0.985.0
+ */
+public class DataflowAnalysisTest {
+
+    @Test
```
Comment (NatashaWso2) [inline tests/ballerina-unit-test/src/test/java/org/ballerinalang/test/dataflow/analysis/DataflowAnalysisTest.java]: Shall we add a description for the test?

**Comment by SupunS** [comment]
Closing this PR, since this was merged to the error-refactor branch

**Pull request merged**